### DISCUSSION TITLE
How migrate from Mycroft on rPi to OVOS

### DISCUSSION THREAD

**Discussion by geetee24** [OpenVoiceOS/ovos-buildroot#182] [Q&A] [answered]
please post url to migration guide and setup on rPi guide.

**Comment by JarbasAl** [comment]
we do not have a migration guide, but we have a repository for creating ovos rpi images

https://github.com/OpenVoiceOS/raspbian-ovos

see the [manual_user_install.sh](https://github.com/OpenVoiceOS/raspbian-ovos/blob/dev/manual_user_install.sh) script if you want to install ovos on an existing raspbian image

**Comment by geetee24** [comment]
do you mean i can install ovos on top of my raspbain that has mycroft on it and all my mycroft scripts will still automatically work?

if not, then how do i copy my existing mycroft intents over to ovos?

**Reply by builderjer** [reply]
The `manual_user_install.sh` script is meant for a fresh install of raspbian.  What mycroft scripts are you referring to?  If it is `mycroft-start` or the like, they are not supported anymore by OVOS, and if you use this script, you have the option of installing and starting systemd files for each process.

If you have custom skills installed, or original Mycroft skills, they can be converted and one contributor even made a script to help the process.
https://github.com/mikejgray/ovos-skill-projen

**Comment by geetee24** [comment]
i wrote many custom skills for mycroft.   so are you saying OVOS can execute them and I cant just copy them over????

**Reply by builderjer** [reply]
Theoretically your skill should work, but many of the properties and functions will be depreciated in version 0.1.0 of ovos-core. There are new best practices and skills should be `pip` installable. The above script will help you convert your skill to ovos best practices. 

Check [this discussion](https://github.com/orgs/OpenVoiceOS/discussions/168) for more

**Comment by geetee24** [comment]
oye.  my skills are not in the public domain and are written on the mycroft so how do they be pip installable since they will reside on the OVOS device?  This is turning into a headache.   I just want my skills to work out of the box :(

**Reply by builderjer** [reply]
Did you try them? What isn't working?

**Reply by builderjer** [reply]
You should be able to clone them to `~/.local/share/mycroft/skills/<skill_name>` and ovos should load it or attempt to

**Comment by geetee24** [comment]
havent tried it yet.  you said that even if it does work, then they will stop working in 0.10 :(   so I have to rewrite them ALL in the new format :(

**Reply by builderjer** [reply]
That script above will do most of the process for you, and they still can be private.

**Reply by builderjer** [reply]
They DO NOT need to be in the public domain for pip to install it. You can install local packages also, if that is your concern

**Comment by geetee24** [comment]
i cant clone them for I cant check them out for they dont reside in a repo so they are not pip installable.   I can file copy them

**Reply by builderjer** [reply]
I think as long ad the file system of the skill works with mycroft-core, that script will work.  @mikejgray would know more. He wrote it

**Reply by builderjer** [reply]
You can place your skill directory into that location also. It does mot need `git clone`

**Reply by geetee24** [reply]
how do i see if ovos ingests my skills from MC  correctly?  Where is a console real time log?

**Reply by builderjer** [reply]
you can use `tail`
`tail -f ~/.local/state/mycroft/phal.log`
or on raspOVOS `ologs` will give you the tail from all of the logs except `bus`

**Reply by geetee24** [reply]
simply copyiing files into the folder DOES nothing.  OVOS doesnt see them.

and the import tool fails too.

Running
projen new ovosskill --from https://github.com/asasasas/asasasas.git --retrofit

results in:

installing external module https://github.com/asasasas/asasasas.git...
Error: Command failed: npm install --save --save-dev -f --no-package-lock --prefix="/home/ovos/.local/share/mycroft/skills" https://github.com/aaa/asasasasas.git
npm WARN using --force Recommended protections disabled.
npm ERR! code ENOENT
npm ERR! syscall open
npm ERR! path /home/ovos/.npm/_cacache/tmp/git-clonejemoST/package.json
npm ERR! errno -2
npm ERR! enoent ENOENT: no such file or directory, open '/home/ovos/.npm/_cacache/tmp/git-clonejemoST/package.json'
npm ERR! enoent This is related to npm not being able to find a file.
npm ERR! enoent

**Reply by mikejgray** [reply]
That's not the right command.

**Reply by geetee24** [reply]
it states the command here: https://github.com/mikejgray/ovos-skill-projen

Retrofitting Mycroft skills (WIP, please open issues if it doesn't work for your use case)
If you have an existing Mycroft skill that you'd like to convert to an OVOS skill, you can do so by running the following command in your skill directory:

projen new ovosskill --from "@mikejgray/ovos-skill-projen@latest" --retrofit

-----------------------

so what is the right command please?

**Reply by mikejgray** [reply]
The right command is the one you just copied from the README. The one you said that you ran above is not the same.

**Reply by geetee24** [reply]
what folder do i run it in?

running in ovos@raspOvos:~/.local/share/mycroft/skills/myskill  fails with EACCESS

I copied my skill into it.  and his tool fails

ovos@raspOvos:~/.local/share/mycroft/skills/aaa $ 
ovos@raspOvos:~/.local/share/mycroft/skills/aaa $ projen new ovosskill --from "@mikejgray/ovos-skill-projen@latest" --retrofit
/home/ovos/.local/share/mycroft/skills/aaa
+-- (empty)

Error: Command failed: npm ls --prefix="/home/ovos/.local/share/mycroft/skills/aaa" --depth=0 --pattern projen || npm install --save --save-dev -f --no-package-lock --prefix="/home/ovos/.local/share/mycroft/skills/aaa" projen@latest
npm WARN using --force Recommended protections disabled.
npm ERR! code EACCES
npm ERR! syscall mkdir
npm ERR! path /home/ovos/.local/share/mycroft/skills/aaa/node_modules
npm ERR! errno -13
npm ERR! Error: EACCES: permission denied, mkdir '/home/ovos/.local/share/mycroft/skills/aaa/node_modules'
npm ERR!  [Error: EACCES: permission denied, mkdir '/home/ovos/.local/share/mycroft/skills/aaa/node_modules'] {
npm ERR!   errno: -13,
npm ERR!   code: 'EACCES',
npm ERR!   syscall: 'mkdir',
npm ERR!   path: '/home/ovos/.local/share/mycroft/skills/aaaa/node_modules'

**Reply by builderjer** [reply]
Is your venv activated?

**Reply by mikejgray** [reply]
Run it in the skill directory itself

**Comment by geetee24** [comment]
no.

**Comment by geetee24** [comment]
do i run it: ~/.local/share/mycroft/skills/mynewskill

or at:  ~/.local/share/mycroft/skills


now i get:

  run `npm fund` for details

found 0 vulnerabilities

👾 Project definition file was created at
/home/ovos/.local/share/mycroft/skills/aaa/.projenrc.json
vocab folder not found in original skill; skipping.
dialog folder not found in original skill; skipping.
regex folder not found in original skill; skipping.
intents folder not found in original skill; skipping.
Found manifest.yml, trying to extract Python dependencies
***@***.***/ovos-skill-projen/lib/index.js:85
            if (manifestObject.dependencies.python) {
                               ^

TypeError: Cannot read properties of null (reading 'dependencies')
    at new OVOSSkillProject
***@***.***/ovos-skill-projen/lib/index.js:85:32)
    at evalmachine.<anonymous>:1:27
    at Script.runInContext (node:vm:141:12)

**Reply by geetee24** [reply]
i tried in the skills folder and it fails with message above.   SO frustrating.

**Reply by geetee24** [reply]
TypeError: 'NoneType' object is not callable
2024-03-01 13:00:35.369 - skills - ovos_core.skill_manager:_unload_skill:603 - INFO - removing dddd
2024-03-01 13:00:53.789 - skills - ovos_workshop.skill_launcher:load:282 - INFO - ATTEMPTING TO LOAD SKILL: mynewskill
2024-03-01 13:00:53.791 - skills - ovos_workshop.skill_launcher:_load_skill_source:421 - ERROR - Failed to load mynewskill due to a missing file.
2024-03-01 13:00:53.793 - skills - ovos_workshop.skill_launcher:_create_skill_instance:473 - WARNING - Legacy skill: 'NoneType' object is not callable
2024-03-01 13:00:53.795 - skills - ovos_core.skill_manager:_load_skill:593 - ERROR - Load of skill /home/ovos/.local/share/mycroft/skills/mynewskill failed!
Traceback (most recent call last):
  File "/home/ovos/.venv/lib/python3.11/site-packages/ovos_workshop/skill_launcher.py", line 469, in _create_skill_instance
    self.instance = skill_creator(bus=self.bus,
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: 'NoneType' object is not callable

During handling of the above exception, another exception occurred:

**Reply by geetee24** [reply]
Traceback (most recent call last):
  File "/home/ovos/.venv/lib/python3.11/site-packages/ovos_workshop/skill_launcher.py", line 424, in _load_skill_source
    skill_module = load_skill_module(main_file_path, self.skill_id)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ovos/.venv/lib/python3.11/site-packages/ovos_workshop/skill_launcher.py", line 82, in load_skill_module
    spec.loader.exec_module(mod)
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/home/ovos/.local/share/mycroft/skills/aaa/__init__.py", line 4, in <module>
    from ovos_utils.intents import IntentBuilder
ModuleNotFoundError: No module named 'ovos_utils.intents'

**Reply by geetee24** [reply]
do i need to remove these?


# TODO: Remove unused OVOS imports
from ovos_workshop.decorators import intent_handler
from ovos_workshop.skills import OVOSSkill
from ovos_utils.intents import IntentBuilder
from ovos_bus_client.message import Message
# TODO: Remove unused OVOS imports
from ovos_workshop.decorators import intent_handler
from ovos_workshop.skills import OVOSSkill
from ovos_utils.intents import IntentBuilder
from ovos_bus_client.message import Message

**Reply by mikejgray** [reply]
What does it say in the comment?

**Comment by JarbasAl** [comment, marked as answer]
To answer the original question
- we dont currently have a migration guide
- we do not support upgrading directly from old mycroft to ovos, a fresh install is necessary
- keeping a existing picroft alive is not officially supported, but was answered here https://github.com/orgs/OpenVoiceOS/discussions/217
- running old Mycroft skills under OVOS is answered here https://github.com/orgs/OpenVoiceOS/discussions/225
- if you want to find official replacements for old skills see a list here https://github.com/OpenVoiceOS/ovos-core/issues/382
- if you want to modernize your skills, docs are at [openvoiceos.github.io/ovos-technical-manual](https://openvoiceos.github.io/ovos-technical-manual/), and https://github.com/mikejgray/ovos-skill-projen does most of the work for you
- official replacement for picroft images is https://github.com/OpenVoiceOS/raspOVOS
- to install OVOS from scratch use https://github.com/OpenVoiceOS/ovos-installer

Let's please keep things separated. 1 question per topic. This thread is from Jul 2023 already and not really on topic anymore. In case of bug reports (not questions, known bugs) please report those in the github issues, in case of further questions please open a new thread, by keeping them separated and with marked answers it makes it much easier for the next user to find help if their problem was addressed before
